File: eco-core/core-backend/src/main/java/com/willfp/eco/internal/proxy/EcoProxyFactory.java
package com.willfp.eco.internal.proxy;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.PluginDependent;
import com.willfp.eco.core.proxy.AbstractProxy;
import com.willfp.eco.core.proxy.ProxyConstants;
import com.willfp.eco.core.proxy.exceptions.ProxyError;
import com.willfp.eco.core.proxy.ProxyFactory;
import com.willfp.eco.core.proxy.exceptions.UnsupportedVersionException;
import org.apache.commons.lang.Validate;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.net.URLClassLoader;
import java.util.Arrays;
import java.util.IdentityHashMap;
import java.util.List;
import java.util.Map;

public class EcoProxyFactory extends PluginDependent<EcoPlugin> implements ProxyFactory {
    private static final List<String> SUPPORTED_VERSIONS = Arrays.asList(
            "v1_16_R3",
            "v1_17_R1"
    );

    private final ClassLoader proxyClassLoader;

    private final Map<Class<? extends AbstractProxy>, AbstractProxy> cache = new IdentityHashMap<>();

    public EcoProxyFactory(@NotNull final EcoPlugin plugin) {
        super(plugin);

        Validate.isTrue(
                !plugin.getProxyPackage().equalsIgnoreCase(""),
                "Specified plugin has no proxy support!"
        );

        this.proxyClassLoader = plugin.getClass().getClassLoader();
    }

    @Override
    public <T extends AbstractProxy> @NotNull T getProxy(@NotNull final Class<T> proxyClass) {
        try {
            T cachedProxy = attemptCache(proxyClass);
            if (cachedProxy != null) {
                return cachedProxy;
            }

            String className = this.getPlugin().getProxyPackage() + "." + ProxyConstants.NMS_VERSION + "." + proxyClass.getSimpleName().replace("Proxy", "");
            Class<?> clazz = this.getPlugin().getClass().getClassLoader().loadClass(className);
            Object instance = clazz.getDeclaredConstructor().newInstance();
            if (proxyClass.isAssignableFrom(clazz) && proxyClass.isInstance(instance)) {
                T proxy = proxyClass.cast(instance);
                cache.put(proxyClass, proxy);
                return proxy;
            }
        } catch (Exception e) {
            throwError(e);
        }

        throwError(new IllegalArgumentException());

        throw new RuntimeException("Something went wrong.");
    }

    private void throwError(@Nullable final Exception e) {
        if (e != null) {
            e.printStackTrace();
        }

        if (!SUPPORTED_VERSIONS.contains(ProxyConstants.NMS_VERSION)) {
            throw new UnsupportedVersionException("You're running an unsupported server version: " + ProxyConstants.NMS_VERSION);
        } else {
            throw new ProxyError("Error with proxies - here's a stacktrace. Only god can help you now.");
        }
    }

    private <T extends AbstractProxy> T attemptCache(@NotNull final Class<T> proxyClass) {
        Object proxy = cache.get(proxyClass);
        if (proxy == null) {
            return null;
        }

        if (proxyClass.isInstance(proxy)) {
            return proxyClass.cast(proxy);
        }

        return null;
    }

    public void clean() {
        try {
            if (proxyClassLoader instanceof URLClassLoader urlClassLoader) {
                urlClassLoader.close();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        cache.clear();
    }
}

File: eco-core/core-backend/src/main/kotlin/com/willfp/eco/internal/proxy/EcoProxyFactory.kt
package com.willfp.eco.internal.proxy

import com.willfp.eco.core.EcoPlugin
import com.willfp.eco.core.proxy.ProxyConstants
import com.willfp.eco.core.proxy.ProxyFactory
import com.willfp.eco.core.proxy.exceptions.ProxyError
import com.willfp.eco.core.proxy.exceptions.UnsupportedVersionError
import java.net.URLClassLoader
import java.util.IdentityHashMap

class EcoProxyFactory(
    private val plugin: EcoPlugin
) : ProxyFactory {
    private val proxyClassLoader: ClassLoader = plugin::class.java.classLoader
    private val cache: MutableMap<Class<out Any>, Any> = IdentityHashMap()

    override fun <T : Any> getProxy(proxyClass: Class<T>): T {
        try {
            val cachedProxy: T? = attemptCache(proxyClass)
            if (cachedProxy != null) {
                return cachedProxy
            }

            val className =
                this.plugin.proxyPackage + "." + ProxyConstants.NMS_VERSION + "." + proxyClass.simpleName.replace(
                    "Proxy",
                    ""
                )
            val clazz = this.plugin::class.java.classLoader.loadClass(className)
            val instance = clazz.getDeclaredConstructor().newInstance()
            if (proxyClass.isAssignableFrom(clazz) && proxyClass.isInstance(instance)) {
                val proxy = proxyClass.cast(instance)
                cache[proxyClass] = proxy
                return proxy
            }
        } catch (e: Exception) {
            throw proxyErrorFrom(e)
        }

        throw proxyErrorFrom(IllegalArgumentException("Class doesn't seem to be a proxy."))
    }

    private fun proxyErrorFrom(e: Exception): Throwable {
        plugin.logger.severe("Fatal error with proxies! This plugin can't load.")

        return if (!ProxyConstants.SUPPORTED_VERSIONS.contains(ProxyConstants.NMS_VERSION)) {
            ProxyError(
                "Could not initialize proxy.",
                UnsupportedVersionError()
            )
        } else {
            ProxyError(
                "Could not initialize proxy. Are you running a supported server version?",
                e
            )
        }
    }

    private fun <T : Any> attemptCache(proxyClass: Class<T>): T? {
        val proxy = cache[proxyClass] ?: return null

        if (proxyClass.isInstance(proxy)) {
            return proxyClass.cast(proxy)
        }

        return null
    }

    fun clean() {
        if (proxyClassLoader is URLClassLoader) {
            proxyClassLoader.close()
        }

        cache.clear()
    }
}
